File: formula-java/src/main/java/org/formula/parse/tree/NotTokenMatcher.java
package org.formula.parse.tree;

import java.util.Objects;

public class NotTokenMatcher implements TokenMatcher {
    private final TokenMatcher matcher;

    @Override
    public boolean matches(char c) {
        return !matcher.matches(c);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        NotTokenMatcher that = (NotTokenMatcher) o;
        return Objects.equals(matcher, that.matcher);
    }

    @Override
    public int hashCode() {
        return matcher != null ? matcher.hashCode() : 0;
    }

    public NotTokenMatcher(TokenMatcher matcher) {
        this.matcher = matcher;
    }
}

File: cs/formula-cs/Formula/TokenTree/NotTokenMatcher.cs
namespace Formula.TokenTree;

public class NotTokenMatcher : ITokenMatcher
{
    private readonly ITokenMatcher _matcher;
    
    public bool Matches(char c) {
        return !_matcher.Matches(c);
    }
    
    public override bool Equals(object? o)
    {
        return o is NotTokenMatcher matcher && Equals(matcher);
    }

    private bool Equals(NotTokenMatcher other)
    {
        return _matcher.Equals(other._matcher);
    }

    public override int GetHashCode()
    {
        return _matcher.GetHashCode();
    }

    public NotTokenMatcher(ITokenMatcher matcher) {
        _matcher = matcher;
    }
}